cli/cache/code_cache.rs

```diff
@@ -0,0 +1,346 @@
+  on_failure: CacheFailure::Blackhole,
+};
+
+#[derive(Clone)]
+pub struct CodeCache {
+  inner: CodeCacheInner,
+}
+
+impl CodeCache {
+  pub fn new(db: CacheDB) -> Self {
+    Self {
+      inner: CodeCacheInner::new(db),
+    }
+  }
+
+  fn ensure_ok<T: Default>(res: Result<T, AnyError>) -> T {
+    match res {
+      Ok(x) => x,
+      Err(err) => {
+        // TODO(mmastrac): This behavior was inherited from before the refactoring but it probably makes sense to move it into the cache
+        // at some point.
+        // should never error here, but if it ever does don't fail
+        if cfg!(debug_assertions) {
+          panic!("Error using code cache: {err:#}");
+        } else {
+          log::debug!("Error using code cache: {:#}", err);
+        }
+        T::default()
+      }
+    }
+  }
+
+  pub fn get_sync(
+    &self,
+    specifier: &str,
+    code_cache_type: code_cache::CodeCacheType,
+    source_hash: Option<&str>,
+    source_timestamp: Option<u64>,
+  ) -> Option<Vec<u8>> {
+    Self::ensure_ok(self.inner.get_sync(
+      specifier,
+      code_cache_type,
+      source_hash,
+      source_timestamp,
+    ))
+  }
+
+  pub fn set_sync(
+    &self,
+    specifier: &str,
+    code_cache_type: code_cache::CodeCacheType,
+    source_hash: Option<&str>,
+    source_timestamp: Option<u64>,
+    data: &[u8],
+  ) {
+    Self::ensure_ok(self.inner.set_sync(
+      specifier,
+      code_cache_type,
+      source_hash,
+      source_timestamp,
+      data,
+    ));
+  }
+}
+
+impl code_cache::CodeCache for CodeCache {
+  fn get_sync(
+    &self,
+    specifier: &str,
+    code_cache_type: code_cache::CodeCacheType,
+    source_hash: Option<&str>,
+    source_timestamp: Option<u64>,
+  ) -> Option<Vec<u8>> {
+    self.get_sync(specifier, code_cache_type, source_hash, source_timestamp)
+  }
+
+  fn set_sync(
+    &self,
+    specifier: &str,
+    code_cache_type: code_cache::CodeCacheType,
+    source_hash: Option<&str>,
+    source_timestamp: Option<u64>,
+    data: &[u8],
+  ) {
+    self.set_sync(
+      specifier,
+      code_cache_type,
+      source_hash,
+      source_timestamp,
+      data,
+    );
+  }
+}
+
+#[derive(Clone)]
+struct CodeCacheInner {
+  conn: CacheDB,
+}
+
+impl CodeCacheInner {
+  pub fn new(conn: CacheDB) -> Self {
+    Self { conn }
+  }
+
+  pub fn get_sync(
+    &self,
+    specifier: &str,
+    code_cache_type: code_cache::CodeCacheType,
+    source_hash: Option<&str>,
+    source_timestamp: Option<u64>,
+  ) -> Result<Option<Vec<u8>>, AnyError> {
+    let mut query = "
+      SELECT
+        data
+      FROM
+        codecache
+      WHERE
+        specifier=?1 AND type=?2"
+      .to_string();
+    let mut params: Vec<rusqlite::types::Value> = vec![
+      specifier.to_string().into(),
+      code_cache_type.as_str().to_string().into(),
+    ];
+
+    let mut column_index = 3;
+    if let Some(source_hash) = source_hash {
+      query += &format!(" AND source_hash=?{}", column_index);
```

Comment:
We're doing quite a few string allocations and concatenations in this function for each file that we load. Is it possible to store null timestamps as zero and empty hashes as empty string and just use a `&static str` query string?

Reply:
+1 from me on this now. Would be awesome if we could make a single query here without string concats

Reply:
Will rework this to keep the existing `&'static str` interface.